trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/Boogie2ACSL.java

```diff
@@ -139,7 +224,9 @@ private BacktranslatedExpression translateIdentifierExpression(
 			final var range = getRangeForCType(type);
 			return new BacktranslatedExpression(new ACSLResultExpression(), type, range);
 		} else if (mMapping.hasVar(boogieId, expr.getDeclarationInformation())) {
-			final Pair<String, CType> pair = mMapping.getVar(boogieId, expr.getDeclarationInformation());
+			final Triple<String, CType, Boolean> pair = mMapping.getVar(boogieId, expr.getDeclarationInformation());
+			// TODO Should we do something different if the variable is on-heap?
```

Comment:
Should we?

Reply:
Good question :grin: If some variable `z` is on-heap, the corresponding Boogie variable `~#z~0` actually represents the pointer to `z`, right? So perhaps we should backtranslate it to `&z` ? Or am I missing something?

Reply:
> If some variable `z` is on-heap, the corresponding Boogie variable `~#z~0` actually represents the pointer to `z`, right?

Yes, I think so.

> So perhaps we should backtranslate it to `&z` ? Or am I missing something?

We could do it like that, but this would also influence all of our backtranslation tests that use pointers.
I would rather propose to keep it for now and handle this in `translateArrayAccess` when we support pointer dereferences there.

Reply:
Ok, to put it another way: While the current code is slightly incorrect, it does not currently lead to a bug because we can't backtranslate expressions with pointers anyway. Is that the gist of it?

Then I would leave the translation as-is, but update the TODO accordingly.

I briefly thought about adding `\old(&z) == &z` for on-heap global variables `z` where `~#z~0` is unmodifiable, but I guess that should be an ACSL tautology, so there's not really a point :wink:

Reply:
> Ok, to put it another way: While the current code is slightly incorrect, it does not currently lead to a bug because we can't backtranslate expressions with pointers anyway. Is that the gist of it?

Yes, kind of. Another point is that pointers are boogie structs, so the variable should not occur standalone in the expression, but only as a struct access.

> Then I would leave the translation as-is, but update the TODO accordingly.

👍 

> I briefly thought about adding `\old(&z) == &z` for on-heap global variables `z` where `~#z~0` is unmodifiable, but I guess that should be an ACSL tautology, so there's not really a point 😉

Not to be confused with `&\old(z) == &z`, but we probably don't want this in the contract either (for now) 😄

Reply:
On a related note (but not something for this PR): It is sometimes confusing to read `z={3:0}` in counterexamples after an assignment `z = 10;`. Writing `&z={3:0}` might make things less confusing?

Reply:
Yes, this might be a good idea.

Reply:
> Another point is that pointers are boogie structs, so the variable should not occur standalone in the expression, but only as a struct access.

That is something to be handled in the backtranslator for `BoogiePreprocessor`. Currently we still have `IdentifierExpression`s with identifier `~#z~0!base` etc.